lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -73,6 +74,128 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+                     bool narrowTrailingBits, PatternRewriter &rewriter,
+                     std::function<Value(ArrayRef<Value>)> createOp) {
+  // If the result is never used, no point optimizing this. It will
+  // also complicated error handling in getLowestBitAndHigestBitRequired.
+  auto usersIterator = narrowingCandidate->getUsers();
+  assert(!usersIterator.empty() &&
+         "narrowingCandidate must have at least one use.");
+  assert(narrowingCandidate->getNumResults() == 1 &&
+         "narrowingCandidate must have exactly one result");
+  IntegerType narrowingCandidateType =
+      narrowingCandidate->getResultTypes().front().cast<IntegerType>();
+
+  size_t highestBitRequired;
+  size_t lowestBitRequired;
+  size_t originalOpWidth = narrowingCandidateType.getIntOrFloatBitWidth();
+
+  std::tie(lowestBitRequired, highestBitRequired) =
+      getLowestBitAndHighestBitRequired(narrowingCandidate, narrowTrailingBits,
+                                        originalOpWidth);
+
+  // Give up, because we can't make it narrower than it already is.
+  if (lowestBitRequired == 0 && highestBitRequired == originalOpWidth - 1)
+    return false;
+
+  auto loc = narrowingCandidate->getLoc();
+  size_t narrowedWidth = highestBitRequired - lowestBitRequired + 1;
+  auto narrowedType = rewriter.getIntegerType(narrowedWidth);
+  Value narrowedOperation =
+      createOp(SmallVector<Value>(llvm::map_range(inputs, [&](auto input) {
+        return rewriter.create<ExtractOp>(loc, narrowedType, input,
+                                          lowestBitRequired);
+      })));
+
+  // Replace all the use-site's extract operation with the newly minted narrowed
+  // version.
+  SmallVector<Operation *, 16> users(usersIterator.begin(),
+                                     usersIterator.end());
+  for (auto *user : users) {
+    auto extractUse = cast<ExtractOp>(user);
+    auto oldLowBit = extractUse.lowBit();
+
+    assert(oldLowBit >= lowestBitRequired &&
+           "incorrectly deduced the lowest bit required in usage arguments.");
+
+    auto loc = extractUse.getLoc();
+
+    // rewriter.replaceOp or replaceOpWithNewOp is potentially memory unsafe, as
+    // it requires the rewriting driver to ensure that dangling references in
+    // any internal data structures are properly cleared out.
+    if (narrowedWidth == extractUse.getType().getWidth()) {
+      extractUse.replaceAllUsesWith(narrowedOperation);
+    } else {
+      uint32_t newLowBit = oldLowBit - lowestBitRequired;
+      auto narrowedExtract = rewriter.create<ExtractOp>(
+          loc, extractUse.getType(), narrowedOperation, newLowBit);
+      extractUse.replaceAllUsesWith(narrowedExtract.getOperation());
+    }
+  }
+
+  // It is not necessary to call rewriter.replaceOp(narrowingCandidate,
+  // narrowedOperation) here, since all the use-site now refers to
+  // `narrowedOperation` instead. narrowingCandidate will be eliminated by DCE.
```

Comment:
You can move this comment up to where this happens :-)

Reply:
I reworded this to something this captures both the non-root case and root-case better =)